### PR TITLE
Allow import of multiple signing certs

### PR DESCRIPTION
**Pull request by rist**
Was using wrong syntax and therefore failed

Resolves https://github.com/tuist/tuist/issues/2109

### Short description 📝

Allow import of multiple certificates - before this PR only the first certificate was successfully imported. The command to check if a certificate was already supported used some unsupported/unknown arguments and therefore failed as soon as the keychain was not empty.

### Solution 📦

Use correct syntax to check for the presence of a public key in the keychain. Was unable to find a mechanism to look up private keys (It seems to be possible only by name, but was unable to extract that name from the private key) but found out that when importing the same private key multiple times keychain handles that and only keeps one entry.

### CODE REVIEW

**Comment by rist** [comment]
first step is to fix import of public keys.
Private key import is not working yet

**Comment by rist** [comment]
Looked more into the issue of private keys:
It seems the only workable solution to look up private keys in the keychain is using their friendly name. In my first use cases the friendly name of the key (e.g. "Some Organisation") was basically a substring of the certificate name ("Apple Development: Some Organisation (1234)". But in my real life use case (7 different subsidiaries managing their AppStore account and certificates) I ran into cases where the friendly name of a private key was the name of the account manager.
I also tried to extract the friendly name of the p12 file but digging into to that using the `openssl` command I never found that friendly name (which in fact is visible in Keychain Assistant).

But then I found a much easier solution - it seems Keychain Access in fact handles duplicate imports of keys and either ignores them or replaces the old instance with a new one.

One last thing was that I had one issue where the name extracted from the `Subject` of the public key had some encoding issues (`\\xC3\\xA4` instead of `ä`) so I added a helper method cleaning this up. No idea how widespread this problem is but I had one case.

File: Sources/TuistSigning/Certificate/CertificateParser.swift [hunk 1]
```diff
@@ -91,3 +91,20 @@ final class CertificateParser: CertificateParsing {
         try System.shared.capture("openssl", "x509", "-inform", "der", "-in", path.pathString, "-noout", "-subject")
     }
 }
+
+private extension String {
+    func sanitizeEncoding() -> String {
+        // we had certificates where encoding in the name was broken - e.g. \\xC3\\xA4 instead of ä
```
Comment (fortmarek) [inline Sources/TuistSigning/Certificate/CertificateParser.swift:97]: I'd rephrase it to be more generally written (no-one will know who `we` is in a few months)

Also, are we sure the following regex is bulletproof? Have you tried to look into why we even get the bad encoding in the first place?

File: Sources/TuistSigning/SecurityController.swift [hunk 2]
```diff
@@ -16,13 +16,16 @@ final class SecurityController: SecurityControlling {
...
             try importToKeychain(at: certificate.publicKey, keychainPath: keychainPath)
             logger.debug("Imported certificate at \(certificate.publicKey.pathString)")
         }
-        if try !keyExists(at: certificate.privateKey, keychainPath: keychainPath) {
+        do {
+            // found no way to check for the presence of a private key in the keychain, but fortunately keychain takes care of duplicate private keys on its own
             try importToKeychain(at: certificate.privateKey, keychainPath: keychainPath)
             logger.debug("Imported certificate private key at \(certificate.privateKey.pathString)")
+        } catch {
+            logger.warning("Importing certificate private key at \(certificate.privateKey.pathString) failed")
```
Comment (fortmarek) [inline Sources/TuistSigning/SecurityController.swift:28]: I'd fail here - also why have you decided not to bubble up error from `importToKeychain`?

If we decide to throw a new error, I'd try to inspect the error we receive and present a better error message when we can, so user can act on that error (in this case it'd be hard for them to tell what exactly went wrong)

File: Sources/TuistSigning/Certificate/CertificateParser.swift (hunk 1 above)
Comment (fortmarek) [reply Sources/TuistSigning/Certificate/CertificateParser.swift:97]: If we decide to keep this function, I'd add some test case for this.

File: Sources/TuistSigning/SecurityController.swift [hunk 3]
```diff
@@ -43,21 +46,11 @@ final class SecurityController: SecurityControlling {
...
-
-    private func certificateExists(at path: AbsolutePath, keychainPath: AbsolutePath) throws -> Bool {
+    private func certificateExists(_ certificate: Certificate, keychainPath: AbsolutePath) throws -> Bool {
         do {
-            try System.shared.run("/usr/bin/security", "find-certificate", path.pathString, "-P", "", "-k", keychainPath.pathString)
-            logger.debug("Skipping importing certificate at \(path.pathString) because it is already present")
-            return true
+            let existingCertificates = try System.shared.capture("/usr/bin/security", "find-certificate", "-c", certificate.name, "-a", keychainPath.pathString)
+            logger.debug("Skipping importing certificate at \(certificate.publicKey.pathString) because it is already present")
+            return !existingCertificates.isEmpty
```
Comment (fortmarek) [inline Sources/TuistSigning/SecurityController.swift:53]: if `existingCertificates.isEmpty == true` then we'd show `Skipping importing certificate` when we should not ... ?

File: Tests/TuistSigningIntegrationTests/SecurityControllerIntegrationTests.swift [hunk 4]
```diff
@@ -57,17 +57,12 @@ final class SecurityControllerIntegrationTests: TuistTestCase {
         let keychainPath = try temporaryPath().appending(component: Constants.DerivedDirectory.signingKeychain)
 
         let currentDirectory = AbsolutePath(#file.replacingOccurrences(of: "file://", with: "")).removingLastComponent()
-        let publicKey = currentDirectory.appending(component: "Target.Debug.cer")
-        let privateKey = currentDirectory.appending(component: "Target.Debug.p12")
+        let certificateParser = CertificateParser()
+        let certificate = try! certificateParser.parse(publicKey: currentDirectory.appending(component: "Target.Debug.cer"), privateKey: currentDirectory.appending(component: "Target.Debug.p12"))
```
Comment (fortmarek) [inline Tests/TuistSigningIntegrationTests/SecurityControllerIntegrationTests.swift:61]: I'd drop the exclamation mark, there is no need to do so here.

File: features/generate-1.feature [hunk 5]
```diff
@@ -58,4 +58,10 @@ Feature: Generate a new project using Tuist (suite 1)
     Then the scheme SignApp has a build setting CODE_SIGN_IDENTITY with value Apple Development: Marek Fort (54GSF6G47V) for the configuration Release
     Then the scheme SignApp has a build setting PROVISIONING_PROFILE_SPECIFIER with value 76a7d75c-01d4-4c7f-9140-1d829227883a for the configuration Release
     Then the scheme SignApp has a build setting PRODUCT_BUNDLE_IDENTIFIER with value team.io.tuist.debug for the configuration Debug
-    Then the scheme SignApp has a build setting PRODUCT_BUNDLE_IDENTIFIER with value team.io.tuist.release for the configuration Release
\ No newline at end of file
+    Then the scheme SignApp has a build setting PRODUCT_BUNDLE_IDENTIFIER with value team.io.tuist.release for the configuration Release
+    Then the scheme AppA has a build setting CODE_SIGN_IDENTITY with value iPhone Developer: Christoph Lederer (F2X7FM3XMV) for the configuration Debug
```
Comment (fortmarek) [inline features/generate-1.feature:62]: this is great, thanks for extending the fixture 👏

**Review by fortmarek — COMMENTED** [review]
PR looks great, left just some minor comments 🎉

**Review by pepibumur — APPROVED** [review]

File: Sources/TuistSigning/Certificate/CertificateParser.swift (hunk 1 above)
Comment (rist) [reply Sources/TuistSigning/Certificate/CertificateParser.swift:97]: In hat specific case it seems that there is already some encoding mistake in the DB of Apple.
When opening that team in AppStoreConnect the `ä` in the company's name is _broken_. Funnily it looks ok in developer.apple.com and the AppStore.

Fixing that would involve a big financial institution and Apple - so I'm not that optimistic about being able to fix the root cause.

But, yes will add tests

**Comment by fortmarek** [comment]
I'd wait for the main branch to be green (although it is not something you can fix in this PR), but otherwise no more comments, thanks!

**Review by fortmarek — APPROVED** [review]
